core/src/main/java/tachyon/master/BlockInfo.java

```diff
@@ -144,13 +146,34 @@ public synchronized InodeFile getInodeFile() {
 
   /**
    * Get the locations of the block, which are the workers' net address who has the data of the
-   * block in memory.
+   * block in its hierarchy store. The list is sorted by the storage level alias(MEM, SSD, HDD).
+   * That is, the worker who has the data of the block in its memory is in the top of the list.
+   * The highest tier locations are shuffled to spread the load.
    * 
    * @return the net addresses of the locations
    */
   public synchronized List<NetAddress> getLocations(TachyonConf tachyonConf) {
-    List<NetAddress> ret = new ArrayList<NetAddress>(mLocations.size());
-    ret.addAll(mLocations.values());
+    NetAddress[] addresses = new NetAddress[mLocations.size()];
+    if (mLocations.size() > 0) {
+      int[] index = new int[StorageLevelAlias.SIZE];
+      for (long storageDirId : mStorageDirIds.values()) {
+        int storageLevelValue = StorageDirId.getStorageLevelAliasValue(storageDirId);
+        index[storageLevelValue - 1] ++;
+      }
+      int highestTierLocNum = index[0];
+      for (int i = 1; i < StorageLevelAlias.SIZE; i ++) {
+        if (highestTierLocNum == 0) {
+          highestTierLocNum = index[i];
+        }
+        index[i] += index[i - 1];
+      }
+      for (Map.Entry<NetAddress, Long> entry : mStorageDirIds.entrySet()) {
+        int storageLevelValue = StorageDirId.getStorageLevelAliasValue(entry.getValue());
+        addresses[-- index[storageLevelValue - 1]] = entry.getKey();
+      }
+      shuffleLocations(addresses, highestTierLocNum);
+    }
```

Comment:
I have a simple way which scans the mStorageDirIds map three times, and adds the MEM locations, SSH locations and HDD locations respectively to the target list in each scan. This could be a little slower but not bad since there are only 3 storage level alias values. Would you prefer to this way?